pylops_mpi/Distributed.py

```diff
@@ -0,0 +1,114 @@
+from pylops_mpi.utils import deps
+
+cupy_message = pylops_deps.cupy_import("the DistributedArray module")
+nccl_message = deps.nccl_import("the DistributedArray module")
+
+if nccl_message is None and cupy_message is None:
+    from pylops_mpi.utils._nccl import (
+        nccl_allgather, nccl_allreduce, nccl_bcast, nccl_send, nccl_recv
+    )
+
+
+class DistributedMixIn:
+    r"""Distributed Mixin class
+
+    This class implements all methods associated with communication primitives
+    from MPI and NCCL. It is mostly charged to identifying which commuicator
+    to use and whether the buffered or object MPI primitives should be used
+    (the former in the case of NumPy arrays or CuPy arrays when a CUDA-Aware
+    MPI installation is available, the latter with CuPy arrays when a CUDA-Aware
+    MPI installation is not available).
+    """
+    def _allreduce(self, base_comm, base_comm_nccl,
+                   send_buf, recv_buf=None, op: MPI.Op = MPI.SUM,
+                   engine="numpy"):
+        """Allreduce operation
+        """
+        if deps.nccl_enabled and base_comm_nccl is not None:
+            return nccl_allreduce(base_comm_nccl, send_buf, recv_buf, op)
+        else:
+            return mpi_allreduce(base_comm, send_buf,
+                                 recv_buf, engine, op)
+
+    def _allreduce_subcomm(self, sub_comm, base_comm_nccl,
+                           send_buf, recv_buf=None, op: MPI.Op = MPI.SUM,
+                           engine="numpy"):
+        """Allreduce operation with subcommunicator
+        """
+        if deps.nccl_enabled and base_comm_nccl is not None:
+            return nccl_allreduce(sub_comm, send_buf, recv_buf, op)
+        else:
+            return mpi_allreduce(sub_comm, send_buf,
+                                 recv_buf, engine, op)
+
+    def _allgather(self, base_comm, base_comm_nccl,
+                   send_buf, recv_buf=None,
+                   engine="numpy"):
+        """Allgather operation
+        """
+        if deps.nccl_enabled and base_comm_nccl is not None:
+            if isinstance(send_buf, (tuple, list, int)):
+                return nccl_allgather(base_comm_nccl, send_buf, recv_buf)
+            else:
+                send_shapes = base_comm.allgather(send_buf.shape)
+                (padded_send, padded_recv) = _prepare_allgather_inputs(send_buf, send_shapes, engine="cupy")
+                raw_recv = nccl_allgather(base_comm_nccl, padded_send, recv_buf if recv_buf else padded_recv)
+                return _unroll_allgather_recv(raw_recv, padded_send.shape, send_shapes)
+        else:
+            if isinstance(send_buf, (tuple, list, int)):
+                return base_comm.allgather(send_buf)
+            return mpi_allgather(base_comm, send_buf, recv_buf, engine)
+
+    def _allgather_subcomm(self, send_buf, recv_buf=None):
```

Comment:
This still needs to be modified like `_allgather` to avoid using `self` inside..